lib/graphql/lang/visitor.ex

```diff
@@ -0,0 +1,184 @@
+      true                               -> {nil, nil}
+    end
+
+    path = cond do
+      parent && !is_nil(item) -> stack.path ++ [key]
+      true                    -> stack.path
+    end
+
+    {item, %Stack{stack | key: key, path: path}}
+  end
+
+  defp leave_node(_stack = %Stack{previous: nil}), do: {nil, nil}
+  defp leave_node(stack) do
+    %{ancestors: ancestors} = stack
+
+    {parent, ancestors} = cond do
+      length(ancestors) === 0 -> {nil, []}
+      true                    -> {List.last(ancestors), Enum.drop(ancestors, -1)}
+    end
+
+    {stack.parent, %Stack{stack |
+      key: List.last(stack.path),
+      path: Enum.drop(stack.path, -1),
+      parent: parent,
+      ancestors: ancestors,
+      index: stack.previous.index,
+      keys: stack.previous.keys,
+      in_list: stack.previous.in_list,
+      previous: stack.previous.previous
+    }}
+  end
+
+  defp apply_visitors(nil, _, _, _), do: nil
+  defp apply_visitors(item, is_leaving, stack, visitors) do
+    cond do
+      not is_list(item) and not is_item(item) -> throw "Invalid AST Node: #{inspect(item)}"
+      is_list(item) -> nil
+      true ->
+        case get_visitor(visitors, item.kind, is_leaving) do
+          {type, visitor} ->
+            args = stack |> Map.take([:key, :parent, :path, :ancestors]) |> Map.merge(%{item: item})
+            case visitor.(args) do
+              %{item: action} -> edit(type, action, item)
+              _               -> nil
+            end
+          nil -> nil
+        end
+    end
+  end
+
+  defp get_visitor(visitors, kind, true),  do: get_visitor(visitors, kind, :leave)
+  defp get_visitor(visitors, kind, false), do: get_visitor(visitors, kind, :enter)
+  defp get_visitor(visitors, kind, type) do
+    # TODO: this is pretty gross
+    cond do
+      Map.has_key?(visitors, kind) ->
+        name_key = Map.get(visitors, kind)
+        cond do
+          is_map(name_key)                            -> {type, Map.get(name_key, type)}  # %{Kind: type: fn()}
+          is_function(name_key, 1) and type == :enter -> {type, name_key} # %{Kind: fn()}
+          true -> nil
+        end
+      Map.has_key?(visitors, type) -> {type, get_in(visitors, [type])} # %{type: fn()}
+      true -> nil
+    end
+  end
+
+  defp is_item(item) do
+    is_map(item) and Map.has_key?(item, :kind) and is_atom(item.kind)
+  end
+
+  defp edit(type, action, item) when is_atom(type) and is_map(item), do: edit(type, action, item)
+  defp edit(:enter, :skip, _item), do: nil                                 # don't visit this item
+  defp edit(:enter, :delete, _item), do: nil                               # delete the item
+  defp edit(:enter, replacement, _item) when is_map(replacement), do: nil  # replace the item
+  defp edit(:leave, :delete, _item), do: nil                               # delete the item
+  defp edit(:leave, replacement, _item) when is_map(replacement), do: nil  # replace the item
```

Comment:
Editing is not implemented yet (and I’m not sure we actually need it?) — will remove this.